### PR TITLE
Always show the last closed report action

### PR DESCRIPTION
**Pull request by mountiny**
### Details

Usually there is only one `closed` report action in a report, however, in case of merging two accounts, when the old account has some archive policy expense chats, there is another closed report action added to the report. The old once (in this scenario - `user removed from policy`) stays in the report so we need to make sure the reportAction we show in the report footer is the `closed` report action with the latest `created` field.

In this PR have added a new method which will filter the report actions for only those which are closed, sorts them and shows the latest one meaning we will always display the most up-to-date archive reason.

### Fixed Issues

$ https://github.com/Expensify/App/issues/14292  

### Tests

- [x] Verify that no errors appear in the JS console

1. Create the first account and create a workspace
2. Invite a second and third account as members to the workspace
3. Send a message in the workspace chat for the second account and the third
accout
4. Remove the second account from the workspace
5. Validate the second account on OldDot
6. Sign in as the third account on OldDot and merge the second account
7. Sign-in to newDot as the third account
8. Verify the archived reason in the archived workspace chat is due to account being merged (and not due to user being removed from the policy)
image

### Offline tests

N/A

### QA Steps

Same as test. Will be queued internally by @trjExpensify 

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

image

</details>

<details>
<summary>Mobile Web - Chrome</summary>

image

</details>

<details>
<summary>Mobile Web - Safari</summary>

image

</details>

<details>
<summary>Desktop</summary>

image

</details>

<details>
<summary>iOS</summary>

image

</details>

<details>
<summary>Android</summary>

image

</details>

### CODE REVIEW

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/libs/ReportActionsUtils.js [hunk 1]
```diff
@@ -179,7 +176,19 @@ function getSortedReportActionsForDisplay(reportActions) {
...
+/**
+ * This method returns the report actions that are ready for display in the ReportActionsView.
+ * The report actions need to be sorted by created timestamp first, and reportActionID second
+ * to ensure they will always be displayed in the same order (in case multiple actions have the same timestamp).
+ * This is all handled with getSortedReportActions() which is used by several other methods to keep the code DRY.
+ *
+ * @param {Object} reportActions
+ * @returns {Array}
+ */
+function getSortedReportActionsForDisplay(reportActions) {
```
[the remark this answers is not in the document]
Comment (mountiny) [reply src/libs/ReportActionsUtils.js:190]: @parasharrajat can you link the duplication? I dont see anything like this on `ReportActionsView:258`

File: src/libs/ReportActionsUtils.js [hunk 2]
```diff
@@ -199,6 +208,19 @@ function getSortedReportActionsForDisplay(reportActions) {
...
+ * In some cases, there can be multiple closed report actions in a chat report.
+ * This method returns the last closed report action so we can always show the correct archived report reason.
+ *
+ * @param {Object} reportActions
+ * @returns {Object}
+ */
+function getLastClosedReportAction(reportActions) {
+    const filteredReportActions = filterOutDeprecatedReportActions(reportActions);
+    const sortedReportActions = getSortedReportActions(filteredReportActions, true);
+    return _.filter(sortedReportActions, reportAction => reportAction.actionName === CONST.REPORT.ACTIONS.TYPE.CLOSED)[0];
```
[the remark this answers is not in the document]
Comment (mountiny) [reply src/libs/ReportActionsUtils.js]: Updated!

File: src/libs/ReportActionsUtils.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mountiny) [reply src/libs/ReportActionsUtils.js:191]: np

**Comment by neil-marcellini** [comment]
👀

**Comment by neil-marcellini** [comment]
Yeah I have the same questions.

I think we need to merge the second account into the third so that the old
account (second account) has archived policy expense chats due to being removed
from a workspace. I think these steps might be better.

1. Create the first account and create a workspace
2. Invite a second and third account as members to the workspace
3. Send a message in the workspace chat for the second account and the third
accout
4. Remove the second account from the workspace
5. Validate the second account on OldDot
6. Sign in as the third account on OldDot and merge the second account
7. Sign-in to newDot as the second account
8. Verify the archived reason in the archived workspace chat is due to account being merged (and not due to user being removed from the policy)

**Comment by neil-marcellini** [comment]
Hmm now when I go to merge accounts I get an error. I'll look into that.
image

**Comment by neil-marcellini** [comment]
Oh I see, the accounts can't be on the wallet beta as shown in the backend [here](https://github.com/Expensify/Web-Expensify/blob/5668bd98742768c367dfdf4bb9e30d84d1e9cb9b/lib/UserAPI.php#L1903-L1905). I'll comment out that line to proceed.

File: src/libs/ReportActionsUtils.js [hunk 3]
```diff
@@ -156,18 +157,15 @@ function getLastVisibleMessageText(reportID, actionsToMerge = {}) {
...
- * The report actions need to be sorted by created timestamp first, and reportActionID second
- * to ensure they will always be displayed in the same order (in case multiple actions have the same timestamp).
- * This is all handled with getSortedReportActions() which is used by several other methods to keep the code DRY.
+ * A helper method to filter out report actions keyed by sequenceNumbers.
  *
  * @param {Object} reportActions
  * @returns {Array}
  */
-function getSortedReportActionsForDisplay(reportActions) {
+function filterOutDeprecatedReportActions(reportActions) {
```
Comment (neil-marcellini) [inline src/libs/ReportActionsUtils.js:165]: 👍 nice I like that you refactored this into a separate function and gave it a descriptive name.

**Review by neil-marcellini — COMMENTED** [review]
The code tests well using the test steps I posted in a previous comment, and the changes look good to me. After you fix that lint error I think it's good to go. Also please update the testing steps to be more clear.

I'm seeing the correct archive reason.
image

**Comment by mountiny** [comment]
Sorry for the it not being more clear, i havent counted the admin account as first or second, so first was the first non-admin member.

Will updat

**Comment by mountiny** [comment]
@parasharrajat Sorry, I am also not fully aware if anything special is required to test this. I hoped not, given @neil-marcellini has tested this and it worked well I think we can proceed with merging once I fix the lint issue.

Thank you both!

**Comment by mountiny** [comment]
@neil-marcellini @parasharrajat Ready to fill out the reviewer checklist and merge, thanks!

**Comment by trjExpensify** [comment]
>Do I need a domain-based email account or Gmail one will work for it?
>Sorry, I am also not fully aware if anything special is required to test this.

The account needs to be on the `policyExpenseChat` beta to access workspace chats.

**Review by neil-marcellini — APPROVED** [review]

**Comment by mountiny** [comment]
@parasharrajat I think we can skip this testing on all platforms, Neil has confirmed this works fine locally and we will test this thoroughly in Staging with Tom so we will catch any issue there nonetheless.

Can you just fill in the checklist and note Neil tested web and we will test the entire flow in staging?

**Review by parasharrajat — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.


:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by mountiny** [comment]
Thanks @parasharrajat

**Comment by mountiny** [comment]
This is merely updating the ReportFooter logic on how to get the closer report action and does not touch Search page logic. I think we are fine here and it was a one off, removing the Deploy blocker label.

Let me know if you think otherwise.

**Comment by mountiny** [comment]
This is building right now but should be ready to test in an hour and tomorrow @trjExpensify

**Comment by trjExpensify** [comment]
Okay, going to jump into this PR review now.

**Comment by trjExpensify** [comment]
Is this related? Created a workspace and my own policyExpenseChat shows with this OpenReport auth error: 

image

**Comment by trjExpensify** [comment]
And then here from the account I'm setting up to be the old employee when trying to access their workspace chat: 

image

**Comment by trjExpensify** [comment]
So the common denominator is that we're running into this bug trying to access your own workspace chat just fine. For example, this is the admin looking at emp17's report: 

image

**Comment by trjExpensify** [comment]
It's traced back to this PR I think: https://github.com/Expensify/Auth/pull/7576

CC: @jasperhuangg @aldo-expensify

**Comment by trjExpensify** [comment]
Okay, that's as much as I can do here to test this PR given that blocker.

**Pull request merged**